flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/transforms/TimeToSqlTypesTransformation.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.types.inference.transforms;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.inference.TypeTransformation;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.LogicalTypeRoot;
+
+import java.sql.Date;
+import java.sql.Time;
+import java.sql.Timestamp;
+
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.hasRoot;
+
+/**
+ * Type transformation that transforms one data type to a new data type whose
+ * conversion class is {@link java.sql.Timestamp}/{@link java.sql.Time}/{@link java.sql.Date}
+ * if the original data type is TIMESTAMP/TIME/DATE.
+ */
+@Internal
+public class TimeToSqlTypesTransformation implements TypeTransformation {
```

Comment:
We could also make this more modular and take a `Map<LogicalTypeRoot, Class>`. We can still offer a convenient `timeToSqlTypes()` method.